Realm.Shared/Dynamic/MetaRealmObject.cs

```diff
@@ -0,0 +1,241 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace Realms.Dynamic
+{
+    internal class MetaRealmObject : DynamicMetaObject
+    {
+        private readonly Realm _realm;
+        private readonly RealmObject.Metadata _metadata;
+
+        private static readonly FieldInfo RealmObjectRealmField = typeof(RealmObject).GetTypeInfo().GetField("_realm", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+        private static readonly FieldInfo RealmObjectRowHandleField = typeof(RealmObject).GetTypeInfo().GetField("_rowHandle", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+
+        public MetaRealmObject(Expression expression, DynamicRealmObject value)
+            : base(expression, BindingRestrictions.Empty, value)
+        {
+            _realm = value.Realm;
+            _metadata = value.ObjectMetadata;
+        }
+
+        public override DynamicMetaObject BindGetMember(GetMemberBinder binder)
+        {
+            var property = _metadata.Schema.Find(binder.Name);
+            if (property == null)
+            {
+                return base.BindGetMember(binder);
+            }
+
+            var arguments = new List<Expression>
+            {
+                WeakConstant(_metadata.Table),
+                Expression.Field(GetLimitedSelf(), RealmObjectRowHandleField),
+                Expression.Constant(_metadata.ColumnIndices[property.Name])
+            };
+
+            MethodInfo realmObjectOpsMethod = null;
+
+            switch (property.Type)
+            {
+                case Schema.PropertyType.Int:
+                    if (property.IsNullable)
+                        realmObjectOpsMethod = typeof(RealmObjectOps).GetMethod(nameof(RealmObjectOps.GetNullableInt64Value));
```

Comment:
> will actually call BindSetMember 3 times and BindGetMember twice?

I believe so, yes.

Our use-case is for [properties to be retrieved in a loop during migration](https://realm.io/docs/swift/latest/#updating-values) so there should be few callsites. Besides, the compiler might be able to optimize in this scenario, but I'm not sure.

Of course, if the compiler doesn't optimize, this'll be slower even when using the statically-typed API because you'll perform two more stores and retrievals than necessary.

Depending on the code our users write we can always optimize if necessary using [`InfoOf.Fody`](https://github.com/Fody/InfoOf).